### DISCUSSION TITLE
Agnostic (Edge) Runtime support

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/wcc#224] [Ideas]
## Overview
To maximize the potential and viability of `wcc` across as many runtimes as possible, an effort should be undertaken to ensure support across "standards" based serverless and edge runtime environments.  Serverless environments are a little more forgiving, but edge functions typically cannot not use things like Node's `fs` or CommonJS.   However, they do [(plan to) standardize on Web APIs like `fetch`, `Request` and `Response`](https://deno.com/blog/announcing-wintercg).

## Details
Basically need to evaluate and allow `wcc` to be usable outside of just NodeJS.  That will mean validating our dependencies for their module system, and any usage of Node specific APIs like `fs`.  Perhaps this may require a pre-bundle if there is an expectation of ESM only as well, but we will have to see.

I think it would be great to reach
- Lambda@Edge
- Netlify Edge Functions
- Cloudflare Workers
- Vercel

**Comment by thescientist13** [comment]
While `path` and `url` are easy enough to work around, `fs` seems like it would be significantly harder without having to employ some sort of "adapter" pattern / feature detection, since it [doesn't seem like it is possible to use `fetch` with a `file://`](https://github.com/ProjectEvergreen/greenwood/issues/953#issuecomment-1345438259), except in Deno.

**Comment by thescientist13** [comment]
FWIW, now with most runtimes supporting NodeJS APIs out of the box, WCC can run on all the above runtimes and hosting providers.  However, not having to use Node APIs internally would be great, but with all the dependencies we use, not sure it would be entirely feasible.
